sys-kernel/asahi-kernel/asahi-kernel-6.11.6_p2.ebuild

```
@@ -0,0 +1,143 @@
+    https://github.com/AsahiLinux/linux/compare/v${MY_BASE}...${ASAHI_TAG}.patch
+        -> linux-${ASAHI_TAG}.patch
+    https://dev.gentoo.org/~mpagano/dist/genpatches/${GENPATCHES_P}.base.tar.xz
+    https://dev.gentoo.org/~mpagano/dist/genpatches/${GENPATCHES_P}.extras.tar.xz
+    https://github.com/projg2/gentoo-kernel-config/archive/${GENTOO_CONFIG_VER}.tar.gz
+        -> gentoo-kernel-config-${GENTOO_CONFIG_VER}.tar.gz
+	https://${FEDORA_CONFIG_DISTGIT}/kernel.git/plain/kernel-aarch64-16k-fedora.config?id=${FEDORA_CONFIG_SHA1}
+		-> kernel-aarch64-16k-fedora.config-${CONFIG_VER}
+"
+
+S="${WORKDIR}/${MY_P}"
+
+LICENSE="GPL-2"
+SLOT="asahi/${PVR}"
+KEYWORDS="arm64"
+
+IUSE="debug"
+
+# Rust is non-negotiable for the dist kernel
+DEPEND="
+	${DEPEND}
+	dev-util/bindgen
+	debug? ( dev-util/pahole )
+"
+
+PDEPEND="
+	~virtual/dist-kernel-${PV}:asahi
+"
+
+QA_FLAGS_IGNORED="
+	usr/src/linux-.*/scripts/gcc-plugins/.*.so
+	usr/src/linux-.*/vmlinux
+	usr/src/linux-.*/arch/powerpc/kernel/vdso.*/vdso.*.so.dbg
+"
+
+src_prepare() {
+    local PATCHES=(
+        # meh, genpatches have no directory
+        "${WORKDIR}"/*.patch
+        "${DISTDIR}/linux-${ASAHI_TAG}.patch"
+        "${FILESDIR}/${PN}-6.8-config-gentoo-Drop-RANDSTRUCT-from-GENTOO_KERNEL_SEL.patch"
+    )
+	default
+
+	# prepare the default config
+	cp "${DISTDIR}/kernel-aarch64-16k-fedora.config-${CONFIG_VER}" ".config" || die
+
+	# ensure a consistant version across kernel and gentoo
+	# this passes the ${PV}-as-release check in kernel-install_pkg_preinst()
+	# override "-asahi" in localversion.05-asahi with "_pX" to override the
+	# kernel's base varsion to gentoo's ${PV}
+	echo "-p${MY_TAG}" > localversion.05-asahi
+	# use CONFIG_LOCALVERSION to provide "asahi" and "dist" annotations.
+	local myversion="-asahi-dist"
+	echo "CONFIG_LOCALVERSION=\"${myversion}\"" > "${T}"/version.config || die
+	local dist_conf_path="${WORKDIR}/gentoo-kernel-config-${GENTOO_CONFIG_VER}"
+
+	local merge_configs=(
+		"${T}"/version.config
+		"${dist_conf_path}"/base.config
+	)
+	use debug || merge_configs+=(
+		"${dist_conf_path}"/no-debug.config
+		"${FILESDIR}"/linux-6.10_disable_debug_info_btf.config
+	)
+
+	# deselect all non APPLE arm64 ARCHs
+	merge_configs+=(
+		"${FILESDIR}"/linux-6.8_arm64_deselect_non_apple_arch.config
+	)
+	# adjust base config for Apple silicon systems
+	merge_configs+=(
+		"${FILESDIR}"/linux-6.8_arch_apple_overrides.config
+	)
+
+	# amdgpu no longer builds with clang (issue #113)
+	merge_configs+=(
+		"${FILESDIR}"/linux-6.10_drop_amdgpu.config
+	)
+
+	kernel-build_merge_configs "${merge_configs[@]}"
+}
+
+src_install() {
+	# call kernel-build's scr_install
+	kernel-build_src_install
+
+	# symlink installed *.dtbs back into kernel "source" directory
+	for dtb in ${ED}/boot/dtbs/${KV_FULL}/apple/*.dtb; do
+		local basedtb=$(basename ${dtb})
+		dosym -r ${EROOT}/boot/dtbs/${KV_FULL}/apple/${basedtb} ${EROOT}/usr/src/linux-${KV_FULL}/arch/arm64/boot/dts/apple/${basedtb}
```

Using `${EROOT}` in `src_install` isn't allowed, and pkgcheck should warn on that.

Would the following work?
```suggestion
		dosym -r ${dtb##${ED}} /usr/src/linux-${KV_FULL}/arch/arm64/boot/dts/apple/
```

---

I pushed a fix for kernel-build.eclass earlier to install the dtb to /lib/modules instead, this is where upstream kernel-install expects it to be. This might need a similar adjustment.

---

Already fixed in the overlay. Just need to bump this MR :)

---

By the way, what is the original reason for symlinking this back into `/usr/src/......`?  Because we could also easily solve this in the eclass (for all dtb files not just apple) if there is some reason to want those files available there as well.

---

This is an Asahi-specific thing. We have a script which updates our second-stage bootloader m1n1, which involves concatenating the dtbs, m1n1 itself, and u-boot into a single blob. Strictly speaking, the DTs _must_ be the ones from the kernel the user wishes to boot (at least, until certain features stabilise).

We symlink the dtbs back into `/usr/src/linux-${KV_FULL}` so that switching between booted kernels is as simple as `eselect kernel set [some kernel] && update-m1n1 && reboot`. This way, `update-m1n1` only has to look for the dtbs in `/usr/src/linux/`. This has become less of a requirement as drivers and bindings stabilise, but is still useful for switching out development kernels for production ones.

This is really only a requirement for Asahi. Not sure if this is useful for other Devicetree platforms, which variously allow Linux to load the FDT by itself (like a module) or have their platform firmware look for them in a well-known location in the filesystem.